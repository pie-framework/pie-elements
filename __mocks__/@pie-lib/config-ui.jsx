import React from 'react';

const Comp = (name) => (props) => <div compName={name}>{name}</div>;

module.exports = {
  NumberTextField: Comp('NumberTextField'),
  InputRadio: Comp('InputRadio'),
  InputCheckbox: Comp('InputCheckbox'),
  TwoChoice: Comp('TwoChoice'),
  InputContainer: Comp('InputContainer'),
  FeedbackConfig: Comp('FeedbackConfig'),
  Tabs: Comp('Tabs'),
  Checkbox: Comp('Checkbox'),
  FeedbackSelector: Comp('FeedbackSelector'),
  choiceUtils: {
    firstAvailableIndex: jest.fn(() => '1'),
  },
};
