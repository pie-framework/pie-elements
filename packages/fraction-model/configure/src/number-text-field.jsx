import { NumberTextField as NTF } from '@pie-lib/pie-toolbox/config-ui';
import PropTypes from 'prop-types';
import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import cn from 'classnames';

export class NumberTextField extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
  };

  render() {
    const { classes } = this.props;
    const props = { ...this.props, classes: undefined };
    return <NTF {...props} className={cn(classes.textField, props.className)} variant="outlined" />;
  }
}

const styles = (theme) => ({
  textField: {
    marginLeft: theme.spacing.unit,
  },
});

const miniStyles = () => ({
  textField: {
    maxWidth: '120px',
    width: '120px',
    marginTop: '0',
    '& [class^="MuiInputBase-root"]': {
      height: 40,
      fontSize: '0.875rem',
    },
  },
  
});
export const MiniField = withStyles(miniStyles)(NumberTextField);

export default withStyles(styles)(NumberTextField);
