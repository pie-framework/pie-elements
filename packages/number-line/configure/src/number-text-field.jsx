import { NumberTextField as NTF } from '@pie-lib/config-ui';
import PropTypes from 'prop-types';
import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import cn from 'classnames';

export class NumberTextField extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
  };

  render() {
    const { classes } = this.props;
    const props = { ...this.props, classes: undefined };
    return <NTF {...props} className={cn(classes.textField, props.className)} variant="outlined" />;
  }
}

const styles = (theme) => ({
  textField: {
    marginLeft: theme.spacing.unit,
  },
});

const miniStyles = () => ({
  textField: {
    maxWidth: '100px',
  },
});
export const MiniField = withStyles(miniStyles)(NumberTextField);

export default withStyles(styles)(NumberTextField);
