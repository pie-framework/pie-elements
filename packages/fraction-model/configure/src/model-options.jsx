import React from 'react';
import PropTypes from 'prop-types';
import { MiniField } from './number-text-field';
import { withStyles } from '@material-ui/core/styles';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import CardBar from './card-bar';
import { Checkbox } from '@pie-lib/pie-toolbox/config-ui';
import cloneDeep from 'lodash/cloneDeep';

export class ModelOptions extends React.Component {
  static propTypes = {
    model: PropTypes.object.isRequired,
    classes: PropTypes.object,
    onChange: PropTypes.func,
    modelOptions: PropTypes.object.isRequired,
  };

  constructor(props) {
    super(props);
    this.changeMaxModel = this.change.bind(this, 'max');
    this.changePartModel = this.change.bind(this, 'part');
    this.studentConfig = this.change.bind(this, 'student-config');
  }

  /**
   * Function to trigger when DOM elements value change for Number type
   * @param {string} key contains key of element
   * @param {object} event contains event object
   * @param {string} value contains value of DOM element
   */
  change(key, event, value) {
    const { model, onChange } = this.props;
    const oldModel = cloneDeep(model);
    let newModel = cloneDeep(model);
    let showDiag = false;
    if (key === 'max') {
      newModel.maxModelSelected = value;
      showDiag = true;
    } else if (key === 'part') {
      newModel.partsPerModel = value;
      showDiag = true;
    } else if (key === 'student-config') {
      newModel.allowedStudentConfig = value;
      showDiag = false;
    }
    onChange(oldModel, newModel, showDiag);
  }

  /**
   * Function to trigger on change of dropdown value model type
   * @param {string} value contains selection value
   */
  handleSelect = (value) => {
    const { model, onChange } = this.props;
    model.modelTypeSelected = value?.target.value;
    onChange(model, { ...model }, false);
  };

  render() {
    const { model, classes, modelOptions } = this.props;
    const { maxOfModel, partsPerModel, modelTypeChoices } = modelOptions;

    return (
      <div>
        <CardBar header="Configure Fraction Model"></CardBar>
        <br />
        <div className={classes.groupInline}>
          <div className={classes.group}>
            <label className={classes.inputLabel}>Model Type</label>
            <Select className={classes.container2} onChange={this.handleSelect} value={model.modelTypeSelected}>
              {modelTypeChoices.map((choice, index) => (
                <MenuItem key={'item_' + index} value={choice.value}>
                  {choice.label}
                </MenuItem>
              ))}
            </Select>
          </div>
          <div className={classes.group}>
            <label className={classes.inputLabel}>Max # of Models</label>
            <MiniField
              min={maxOfModel.min}
              max={maxOfModel.max}
              value={model.maxModelSelected}
              name="max-model"
              onChange={this.changeMaxModel}
            />
          </div>
          <div className={classes.group}>
            <label className={classes.inputLabel}>Parts per Model</label>
            <MiniField
              min={partsPerModel.min}
              max={partsPerModel.max}
              value={model.partsPerModel}
              name="model-parts"
              onChange={this.changePartModel}
            />
          </div>
        </div>
        <div className={classes.checkbox}>
          <Checkbox onChange={this.studentConfig} checked={model.allowedStudentConfig} label={''} />
          <span className={classes.chkLabel}>Allow student to configure number of models and parts per model</span>
        </div>
      </div>
    );
  }
}

const styles = () => ({
  groupInline: {
    alignItems: 'center',
    display: 'flex',
    gap: '1.25rem',
  },
  group: {
    margin: '0.75rem 0',
  },
  inputLabel: {
    display: 'block',
    marginBottom: '0.25rem',
  },
  labelWidth: {
    width: '35%',
  },
  container2: {
    width: '80px',
    alignItems: 'center',
    height: '40px',
    verticalAlign: 'top',
    border: '1px solid rgb(0, 0, 0, 0.23)',
    borderRadius: '4px',
    marginBottom: '8px',
    paddingLeft: '10px',
    ':hover': {
      border: '1px solid rgb(0, 0, 0, 0.87)',
    },
  },
  checkbox: {
    marginLeft: '-15px',
  },
  chkLabel: {
    verticalAlign: 'middle',
  },
});

export default withStyles(styles, { name: 'ModelOptions' })(ModelOptions);
