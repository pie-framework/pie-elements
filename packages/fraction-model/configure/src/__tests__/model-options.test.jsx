import { shallow } from 'enzyme';
import React from 'react';
import defaultValues from '../defaults';
import { ModelOptions } from '../model-options';
import CardBar from '../card-bar';
import { MiniField } from '../number-text-field';
import Select from '@material-ui/core/Select';
import { Checkbox } from '@pie-lib/pie-toolbox/config-ui';

jest.mock('@pie-lib/pie-toolbox/config-ui', () => ({
  Checkbox: (props) => <div {...props} />,
  layout: {
    ConfigLayout: (props) => <div>{props.children}</div>,
  },
}));

jest.mock('@pie-lib/pie-toolbox/editable-html', () => ({
  EditableHtml: (props) => <div {...props} />,
}));

jest.mock('@material-ui/core', () => ({
  Select: (props) => <div {...props} />,
}));

jest.mock('lodash/debounce', () => (fn) => fn);

export const defaultProps = {
  model: {
    correctResponse: [],
    title: '',
    question: '',
    modelTypeSelected: 'bar',
    maxModelSelected: 1,
    partsPerModel: 5,
    allowedStudentConfig: false,
    showGraphLabels: false,
  },
  modelOptions: defaultValues.configuration.modelOptions,
};

describe('Model Options', () => {
  let wrapper;
  let component;

  beforeEach(() => {
    wrapper = (props) => {
      const configProps = {
        classes: {},
        ...defaultProps,
        ...props,
      };
      return shallow(<ModelOptions {...configProps} />);
    };
  });

  it('snapshot renders', () => {
    const component = wrapper();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly: Components mounted correctly', () => {
    component = wrapper();
    expect(component.find(CardBar).length).toEqual(1);
    expect(component.find(Select).length).toEqual(1);
    expect(component.find(MiniField).length).toEqual(2);
    expect(component.find(Checkbox).length).toEqual(1);
  });

  it('render correctly: update model type to Pie', () => {
    let onChange = jest.fn();

    component = wrapper({
      onChange,
    });

    component.instance().handleSelect({
      target: { value: 'pie' },
    });

    expect(onChange).toBeCalledWith(
      expect.objectContaining({ modelTypeSelected: 'pie' }),
      expect.objectContaining({ modelTypeSelected: 'pie' }),
      false,
    );
  });

  it('render correctly: update model type to Bar', () => {
    let onChange = jest.fn();

    component = wrapper({
      onChange,
    });

    component.instance().handleSelect({
      target: { value: 'bar' },
    });

    expect(onChange).toBeCalledWith(
      expect.objectContaining({ modelTypeSelected: 'bar' }),
      expect.objectContaining({ modelTypeSelected: 'bar' }),
      false,
    );
  });

  it('render correctly: update max no of model', () => {
    let onChange = jest.fn();

    component = wrapper({
      onChange,
    });

    component.instance().change('max', {}, 4);

    expect(onChange).toBeCalledWith(
      expect.objectContaining({ maxModelSelected: defaultProps.model.maxModelSelected }),
      expect.objectContaining({ maxModelSelected: 4 }),
      true,
    );
  });

  it('render correctly: update parts per model', () => {
    let onChange = jest.fn();

    component = wrapper({
      onChange,
    });

    component.instance().change('part', {}, 7);

    expect(onChange).toBeCalledWith(
      expect.objectContaining({ partsPerModel: defaultProps.model.partsPerModel }),
      expect.objectContaining({ partsPerModel: 7 }),
      true,
    );
  });

  it('render correctly: update student config checkbox', () => {
    let onChange = jest.fn();

    component = wrapper({
      onChange,
    });

    component.instance().change('student-config', {}, true);

    expect(onChange).toBeCalledWith(
      expect.objectContaining({ allowedStudentConfig: defaultProps.model.allowedStudentConfig }),
      expect.objectContaining({ allowedStudentConfig: true }),
      false,
    );
  });
});
