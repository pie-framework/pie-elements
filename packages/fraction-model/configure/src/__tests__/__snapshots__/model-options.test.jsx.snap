// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model Options snapshot renders 1`] = `
<div>
  <WithStyles(CardBar)
    header="Configure Fraction Model"
  />
  <br />
  <div>
    <div>
      <label>
        Model Type
      </label>
      <WithStyles(WithFormControlContext(Select))
        onChange={[Function]}
        value="bar"
      >
        <WithStyles(MenuItem)
          key="item_0"
          value="bar"
        >
          Bar
        </WithStyles(MenuItem)>
        <WithStyles(MenuItem)
          key="item_1"
          value="pie"
        >
          Pie
        </WithStyles(MenuItem)>
      </WithStyles(WithFormControlContext(Select))>
    </div>
    <div>
      <label>
        Max # of Models
      </label>
      <WithStyles(NumberTextField)
        max={9}
        min={1}
        name="max-model"
        onChange={[Function]}
        value={1}
      />
    </div>
    <div>
      <label>
        Parts per Model
      </label>
      <WithStyles(NumberTextField)
        max={9}
        min={1}
        name="model-parts"
        onChange={[Function]}
        value={5}
      />
    </div>
  </div>
  <div>
    <Checkbox
      checked={false}
      label=""
      onChange={[Function]}
    />
    <span>
      Allow student to configure number of models and parts per model
    </span>
  </div>
</div>
`;
